File: vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.java
class Solution {
    class Pair {
        TreeNode node;
        int width;
        
        Pair(TreeNode node, int width) {
            this.node = node;
            this.width = width;
        }
    }
    
    int leftWidth = 0, rightWidth = 0;
    
    public void width(TreeNode node, int width) {
        if(node == null)
            return;
            
        leftWidth = Math.min(leftWidth, width);
        rightWidth = Math.max(rightWidth, width);
        
        width(node.left, width - 1);
        width(node.right, width + 1);
        
        return;
    }
    
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        List<List<Integer>> arr = new ArrayList<>();
        
        if(root == null)
            return arr;
            
        width(root, 0);
        int wid = rightWidth - leftWidth + 1;
        
        for(int i = 0; i < wid; i++) 
            arr.add(new ArrayList<>());
        
        // here min priority queue is used because we want the elements of same width and vertical order in sorted form
        // Two queues are used so that the parent and child nodes are seperated at each level to avoid mixing of them as we are applying sorted on one horizontal level at a time
        
        PriorityQueue<Pair> parentQ = new PriorityQueue<>((a, b) -> {
            return a.node.val - b.node.val;
        });
        PriorityQueue<Pair> childQ = new PriorityQueue<>((a, b) -> {
            return a.node.val - b.node.val;
        });
        
        int x = Math.abs(leftWidth);
        parentQ.add(new Pair(root, x));
        
        while(parentQ.size() > 0) {
            int size = parentQ.size();
            
            while(size-- > 0) {
                Pair p = parentQ.remove();
                    
                arr.get(p.width).add(p.node.val);
                
                if(p.node.left != null) 
                    childQ.add(new Pair(p.node.left, p.width - 1));
                    
                if(p.node.right != null) 
                    childQ.add(new Pair(p.node.right, p.width + 1));
            }
            
            PriorityQueue<Pair> temp = parentQ;
            parentQ = childQ;
            childQ = temp;
        }
        
        return arr;
    }
    
    // Time complexity = O(N.logN)
    // Space complexity = O(N) 
}

File: 1029-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        map<int, map<int, multiset<int>>> nodes; 
        queue<pair<TreeNode*, pair<int, int>>> todo; 
        todo.push({root, {0, 0}});
        
        while (!todo.empty()) {
            pair<TreeNode*, pair<int, int>> front = todo.front();
            todo.pop();
            TreeNode* node = front.first;
            int x = front.second.first, y = front.second.second;
            
            nodes[y][x].insert(node->val);
            
            if (node->left) {
                todo.push({node->left, {x + 1, y - 1}});
            }
            if (node->right) {
                todo.push({node->right, {x + 1, y + 1}});
            }
        }
        
        vector<vector<int>> ans;
        for (auto p = nodes.begin(); p != nodes.end(); ++p) { 
            vector<int> col;
            for (auto q = p->second.begin(); q != p->second.end(); ++q) {
                col.insert(col.end(), q->second.begin(), q->second.end());
            }
            ans.push_back(col);
        }
        return ans; 
    }
};